pkg/balancer/hashring_test.go

```diff
@@ -0,0 +1,392 @@
+
+	"github.com/authzed/spicedb/pkg/consistent"
+)
+
+type fakeSubConn struct {
+	balancer.SubConn
+	id string
+}
+
+func (fakeSubConn) Connect() {}
+
+// Note: this is testing picker behavior and not the hashring
+// behavior itself, see `pkg/consistent` for tests of the hashring.
+func TestConsistentHashringPickerPick(t *testing.T) {
+	rnd := rand.New(rand.NewSource(1))
+	tests := []struct {
+		name    string
+		spread  uint8
+		rf      uint16
+		info    balancer.PickInfo
+		want    balancer.PickResult
+		wantErr bool
+	}{
+		{
+			name:   "pick one",
+			spread: 1,
+			rf:     100,
+			info: balancer.PickInfo{
+				Ctx: context.WithValue(context.Background(), CtxKey, []byte("test")),
+			},
+			want: balancer.PickResult{
+				SubConn: &fakeSubConn{id: "1"},
+			},
+		},
+		{
+			name:   "pick another",
+			spread: 1,
+			rf:     100,
+			info: balancer.PickInfo{
+				Ctx: context.WithValue(context.Background(), CtxKey, []byte("test2")),
+			},
+			want: balancer.PickResult{
+				SubConn: &fakeSubConn{id: "3"},
+			},
+		},
+		{
+			name:   "pick with spread",
+			spread: 2,
+			rf:     100,
+			info: balancer.PickInfo{
+				Ctx: context.WithValue(context.Background(), CtxKey, []byte("test")),
+			},
+			want: balancer.PickResult{
+				// without spread, this would always be 1.
+				// it can be 1 or 3 with spread 2, but pinning the seed makes it always 3 in the test
+				SubConn: &fakeSubConn{id: "3"},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			p := &consistentHashringPicker{
+				hashring: consistent.MustNewHashring(xxhash.Sum64, tt.rf),
+				spread:   tt.spread,
+				rand:     rnd,
+			}
+			require.NoError(t, p.hashring.Add(subConnMember{key: "1", SubConn: &fakeSubConn{id: "1"}}))
+			require.NoError(t, p.hashring.Add(subConnMember{key: "2", SubConn: &fakeSubConn{id: "2"}}))
+			require.NoError(t, p.hashring.Add(subConnMember{key: "3", SubConn: &fakeSubConn{id: "3"}}))
+			got, err := p.Pick(tt.info)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Pick() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("Pick() got = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestConsistentHashringBalancerConfigToServiceConfigJSON(t *testing.T) {
+	tests := []struct {
+		name              string
+		replicationFactor uint16
+		spread            uint8
+		want              string
+		wantErr           bool
+	}{
+		{
+			name:              "sets rf and spread",
+			replicationFactor: 300,
+			spread:            2,
+			want:              `{"loadBalancingConfig":[{"consistent-hashring":{"replicationFactor":300,"spread":2}}]}`,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &ConsistentHashringBalancerConfig{
+				ReplicationFactor: tt.replicationFactor,
+				Spread:            tt.spread,
+			}
+			got, err := c.ToServiceConfigJSON()
+			if (err != nil) != tt.wantErr {
```

Comment:
I generally prefer to do 

```if tt.wantErr {
  // ensure error
  return
}

// ensure non-error case
```

Reply:
I can change it; I sometimes use goland's test scaffolding which writes assertions like this, and I don't always change them to more idiomatic `testify`